src/Data/Array.ss

```diff
@@ -34,117 +38,174 @@
 
   (define rangeImpl
     (lambda (start end)
-      (let ([res (srfi:214:make-flexvector (+ (- end start) 1))])
-        (srfi:214:flexvector-map/index! (lambda (i x) (+ i start)) res)
-        res)))
+      (let* ([step (if (> start end) -1 1)]
+             [result (srfi:214:make-flexvector (+ (* step (- end start)) 1))])
+        (let recur ([i start]
+                    [n 0])
+          (if (not (= i end))
+            (begin
+              (srfi:214:flexvector-set! result n i)
+              (recur (+ i step) (+ n 1)))
+            (begin
+              (srfi:214:flexvector-set! result n i)
+              result))))))
 
   (define replicateImpl
     (lambda (count value)
-      (error #f "replicateImpl not implemented")))
-
+      (if (< count 1)
+        (rt:make-array)
+        (let ([result (srfi:214:make-flexvector count)])
+          (srfi:214:flexvector-fill! result value)
+          result))))
+      
   (define fromFoldableImpl
     (lambda (foldr xs)
-      (error #f "fromFoldableImpl not implemented")))
+      (define kons (lambda (head) (lambda (tail) (cons head tail))))
+      (srfi:214:list->flexvector (((foldr kons) '()) xs))))
 
-  (define length
-    (lambda (xs)
-      (error #f "length not implemented")))
+  (define length rt:array-length)
 
   (define unconsImpl
     (lambda (empty next xs)
-      (error #f "unconsImpl not implemented")))
+      (if (= (rt:array-length xs) 0)
+        (empty 'unit)
+        ((next (rt:array-ref xs 0)) (srfi:214:flexvector-copy xs 1)))))
 
   (define indexImpl
     (lambda (just nothing xs i)
-      (error #f "indexImpl not implemented")))
+      (if (or (< i 0) (>= i (rt:array-length xs)))
+        nothing
```

Comment:
Should this be applied, like this?
```suggestion
        (nothing)
```

(`just` is a function, so I assume `nothing` is as well)

Reply:
We are passing in the value of `Nothing`, not the constructor itself. Even so, zero-arity constructors are not functions they are just symbols:

```scheme
  (scm:define Nothing
    (scm:quote Nothing))
```